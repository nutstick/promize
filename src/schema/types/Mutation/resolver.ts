import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
  Mutation: {
  },
};

export default resolver;
