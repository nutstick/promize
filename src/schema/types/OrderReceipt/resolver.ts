import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
  OrderReceipt: {
    async product({ product }, _, { database }) {
      return await database.Product.findOne({ _id: product });
    },
    async creator({ creator }, _, { database }) {
      return await database.User.findOne({ _id: creator });
    },
  },
};

export default resolver;
