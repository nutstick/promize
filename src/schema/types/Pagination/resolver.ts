import { GraphQLError } from 'graphql';
import { toObjectID } from 'iridium';
import { Cursor as MongoCursor } from 'mongodb';

export type Base64String = string;

export function base64(i: string): Base64String {
  return Buffer.from(i, 'utf8').toString('base64');
}

export function unbase64(i: Base64String): string {
  return Buffer.from(i, 'base64').toString('utf8');
}

type Cursor = string;

interface SortField {
  sort: string;
  direction: 'ASC' | 'DESC';
}

export interface PaginationOption {
  name: string;
  first?: number;
  after?: Cursor;
  last?: number;
  before?: Cursor;
  orderBy?: SortField[];
}

export function toCursor(node: any, name: string, orderBy: SortField[]) {
  return base64(JSON.stringify(
    orderBy.reduce((prev, order) => ({
      ...prev,
      [order.sort]: node[order.sort],
    }), { name }),
  ));
}

export async function pagination<T = any>(cursor: MongoCursor<T>, options: PaginationOption) {
  const { after, first, before, last } = options;

  // Add default sort by _id
  const orderBy: SortField[] = options.orderBy ? options.orderBy.concat([{
    sort: '_id', direction: 'ASC',
  }]) : [{ sort: '_id', direction: 'ASC' }];

  // Parse orderBy to mongodb sort format
  const sort: any = orderBy ? orderBy.reduce((prev, order) => ({
    ...prev,
    [order.sort]: (order.direction === 'ASC' ? 1 : -1),
  }), {}) : {};

  if (after) {
    try {
      const { name, _id, ...cursorData } = JSON.parse(unbase64(after));

      if (name !== options.name) {
        throw new GraphQLError(`Type of 'after' is not a ${name}`);
      }
      // FIXME: @types/MongoDB Cursor.min(number)
      cursor = (cursor as any).min({
        ...cursorData,
        ..._id && {
          _id: toObjectID(_id),
        },
      });
    } catch (err) {
      if (err instanceof GraphQLError) {
        throw err;
      } else {
        throw new Error(`Can't parse cursor on field 'after' make use you use Cursor type`);
      }
    }
  }
  if (first) {
    cursor = cursor.limit(first + 1);
  }
  if (before) {
    try {
      const { name, _id, ...cursorData } = JSON.parse(unbase64(before));
      if (name !== options.name) {
        throw new GraphQLError(`Type of 'before' is not ${options.name}`);
      }
      // FIXME: @types/MongoDB Cursor.min(number)
      cursor = (cursor as any).max({
        ...cursorData,
        ..._id && {
          _id: toObjectID(_id),
        },
      });
    } catch (err) {
      if (err instanceof GraphQLError) {
        throw err;
      } else {
        throw new Error(`Can't parse cursor on field 'before' make use you use Cursor type`);
      }
    }
  }
  if (last) {
    cursor = cursor.limit(last + 1);
  }

  cursor = cursor.sort(sort);

  const totalCount = await cursor.count();
  const results = await cursor.toArray();

  if (results.length > (first || last)) {
    const endCursor = toCursor(results[results.length - 1], options.name, orderBy);

    return {
      totalCount,
      edges: results.slice(0, -1).map((node) => ({
        node,
        cursor: toCursor(node, options.name, orderBy),
      })),
      pageInfo: {
        endCursor,
        hasNextPage: true,
      },
    };
  }

  return {
    totalCount,
    edges: results.map((node) => ({
      node,
      cursor: toCursor(node, options.name, orderBy),
    })),
    pageInfo: {
      endCursor: null,
      hasNextPage: false,
    },
  };
}
