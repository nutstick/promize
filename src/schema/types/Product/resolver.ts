import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
  ProductEdges: {
  },
  Product: {
    promotionStart({ promotion_start }) {
      return promotion_start;
    },
    promotionEnd({ promotion_end }) {
      return promotion_end;
    },
    originalPrice({ orginal_price }) {
      return orginal_price;
    },
    async owner({ owner }, _, { database }) {
      return await database.User.findOne(owner);
    },
  },
};

export default resolver;
