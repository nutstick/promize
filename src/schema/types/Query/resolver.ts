import * as BluebirdPromise from 'bluebird';
import * as fs from 'fs';
import { join } from 'path';
import { locales } from '../../../config';
import { IResolver } from '../index';

// A folder with messages
// In development, source dir will be used
const MESSAGES_DIR = process.env.MESSAGES_DIR || join(__dirname, './messages');

const readFile = BluebirdPromise.promisify(fs.readFile);

const resolver: IResolver<any, any> = {
  Query: {
    helloworld() {
      return 'Hello Word';
    },
    async me(_, __, { database, user }) {
      if (user && user._id) {
        return await database.User.findOne({ _id: user._id });
      }
      return null;
    },
    async intl(_, { locale }) {
      if (!locales.includes(locale)) {
        throw new Error(`Locale '${locale}' not supported`);
      }

      let localeData;
      try {
        localeData = await readFile(join(MESSAGES_DIR, `${locale}.json`));
      } catch (err) {
        if (err.code === 'ENOENT') {
          throw new Error(`Locale '${locale}' not found`);
        }
      }

      return JSON.parse(localeData);
    },
  },
};

export default resolver;
