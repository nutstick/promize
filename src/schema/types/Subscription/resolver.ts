import { withFilter } from 'graphql-subscriptions';
import { IMessageDocument } from '../../models/Message';
import { ISubsciption } from '../index';
import { MESSAGE_ADDED, pubsub } from '../pubsub';

const resolver: ISubsciption = {
  Subscription: {
    messageAdded: {
      resolve(message) {
        if (message._id) {
          return message;
        }
        return null;
      },
      subscribe: withFilter(
        () => pubsub.asyncIterator(MESSAGE_ADDED),
        (message: IMessageDocument, args) => {
          return message && message.traderoom === args.traderoom;
        },
      ),
    },
  },
};

export default resolver;
