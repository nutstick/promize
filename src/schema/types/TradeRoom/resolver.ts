import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
};

export default resolver;
