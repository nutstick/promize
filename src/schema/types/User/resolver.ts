import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
  User: {
  },
};

export default resolver;
