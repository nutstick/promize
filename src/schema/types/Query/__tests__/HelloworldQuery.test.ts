import { InMemoryCache } from 'apollo-cache-inmemory';
import ApolloClient from 'apollo-client';
import { graphql } from 'graphql';
import gql from 'graphql-tag';
import { Schema } from '../../../';
import { mongodb } from '../../../../config';
import { ServerLink } from '../../../../core/ServerLink';
import { Database } from '../../../models';
import resolver from '../resolver';

interface Query {
  helloworld: string;
}

const query = gql`query Helloworld {
  helloworld
}`;

const database = new Database({
  ...mongodb,
  database: 'test',
});

beforeEach(() => database.connect());

beforeEach(() => database.close());

describe('GraphQL Query', () => {
  it('helloworld', () => {
    const client = new ApolloClient({
      link: new ServerLink({
        schema: Schema,
      }),
      cache: new InMemoryCache(),
      ssrMode: true,
    });

    return client.query<Query>({
      query,
    })
      .then(({ data }) => {
        // TODO: expect all fields
        // expect(data).toMatchSnapshot();
      });
  });
});
