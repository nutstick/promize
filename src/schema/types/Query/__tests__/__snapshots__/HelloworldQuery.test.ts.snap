// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Query helloworld 1`] = `
Object {
  "helloworld": "Hello World",
  Symbol(id): "ROOT_QUERY",
}
`;
